ocaml/xapi/repository.ml

```diff
@@ -510,39 +529,61 @@ let apply_immediate_guidances ~__context ~host ~guidances =
             (* EvacuateHost and RestartToolstack *)
             Client.Host.restart_agent ~rpc ~session_id ~host
-        | l when eq_set2 l ->
+        | l when GuidanceSet.eq_set2 l ->
             (* RestartDeviceModel and RestartToolstack *)
-            restart_device_models ~__context host ;
+            restart_device_models ~__context ~host ;
             Client.Host.restart_agent ~rpc ~session_id ~host
-        | l when eq_set3 l ->
+        | l when GuidanceSet.eq_set3 l ->
             (* RestartDeviceModel and EvacuateHost *)
             (* Evacuating host restarted device models already *)
-            if num_of_hosts = 1 then restart_device_models ~__context host ;
+            if num_of_hosts = 1 then restart_device_models ~__context ~host ;
             ()
-        | l when eq_set4 l ->
+        | l when GuidanceSet.eq_set4 l ->
             (* EvacuateHost, RestartToolstack and RestartDeviceModel *)
             (* Evacuating host restarted device models already *)
-            if num_of_hosts = 1 then restart_device_models ~__context host ;
+            if num_of_hosts = 1 then restart_device_models ~__context ~host ;
             Client.Host.restart_agent ~rpc ~session_id ~host
         | l ->
-            let ref = Ref.string_of host in
+            let host' = Ref.string_of host in
             error
               "Found wrong guidance(s) after applying updates on host \
                ref='%s': %s"
-              ref
+              host'
               (String.concat ";" (List.map Guidance.to_string l)) ;
-            raise Api_errors.(Server_error (apply_guidance_failed, [ref])))
+            raise Api_errors.(Server_error (apply_guidance_failed, [host'])))
   with e ->
-    let ref = Ref.string_of host in
-    error "applying immediate guidances on host ref='%s' failed: %s" ref
+    let host' = Ref.string_of host in
+    error "applying immediate guidances on host ref='%s' failed: %s" host'
       (ExnHelper.string_of_exn e) ;
-    raise Api_errors.(Server_error (apply_guidance_failed, [ref]))
+    raise Api_errors.(Server_error (apply_guidance_failed, [host']))
+
+let apply_pending_guidances ~__context ~host ~guidances =
+  try
+    let open Guidance in
+    guidances
+    |> List.iter (function
+         | RebootHost ->
+             Db.Host.add_pending_guidances ~__context ~self:host
+               ~value:`RebootHost
+         | RestartToolstack ->
+             Db.Host.add_pending_guidances ~__context ~self:host
+               ~value:`RestartToolstack
+         | RestartDeviceModel ->
+             set_pending_restart_device_models ~__context ~host
+         | g ->
+             warn "Unsupported pending guidance %s, ignore it."
+               (Guidance.to_string g))
+  with e ->
+    let host' = Ref.string_of host in
+    error "applying pending guidances on host ref='%s' failed: %s" host'
+      (ExnHelper.string_of_exn e) ;
+    raise Api_errors.(Server_error (apply_guidance_failed, [host']))
```

Comment:
I don't think this is the right error either. In this case, the functions called inside of this function shouldn't raise any exceptions because it is just DB sets and nothing is actually "applied".